### PR TITLE
Remote templates

### PR DESCRIPTION
**Pull request by ulrikstrid**

### CODE REVIEW

File: npm-cli/.npmignore [hunk 1]
```diff
@@ -1 +1,2 @@
-src/*
\ No newline at end of file
+src/*
+share/*
```
Comment (ulrikstrid) [inline npm-cli/.npmignore]: Should just remove the share folder probably

File: npm-cli/rollup.config.js [hunk 2]
```diff
@@ -9,10 +9,10 @@ export default {
     format: "cjs"
   },
   plugins: [
-    nodeResolve(),
+    nodeResolve({ preferBuiltins: true }),
```
Comment (ulrikstrid) [inline npm-cli/rollup.config.js]: This removes the warning

File: npm-cli/src/Pesy.re [hunk 3]
```diff
@@ -36,188 +24,79 @@ let substituteTemplateValues = s =>
...
-/*       packageNameUpperCamelCase, */
-/*     ), */
-/*     spf( */
-/*       "require: [%s]", */
-/*       String.concat(", ", [spf("\"%s\"", packageLibName)]), */
-/*     ), */
-/*   ], */
-/* ]); */
-/* print_newline(); */
+download_git("github:ulrikstrid/hello-reason-pesy", projectPath, error =>
```
Comment (ulrikstrid) [inline npm-cli/src/Pesy.re]: We should take this as a parameter

File: npm-cli/src/Pesy.re [hunk 4]
```diff
@@ -36,188 +24,79 @@ let substituteTemplateValues = s =>
...
+      },
+    );
 
-let id = Spinner.start("\x1b[2mRunning\x1b[0m esy install");
-exec("esy i", (e, stdout, stderr) => {
-  Spinner.stop(id);
-  if (Js.eqNullable(e, Js.Nullable.null)) {
-    let id = Spinner.start("\x1b[2mRunning\x1b[0m esy pesy\x1b[2m and \x1b[0m building project dependencies");
-    exec("esy pesy", (e, stdout, stderr) => {
+    let libKebab = packageNameKebabSansScope;
```
Comment (ulrikstrid) [inline npm-cli/src/Pesy.re]: Everything from here is basically just refmt.

File: npm-cli/src/Pesy.re [hunk 5]
```diff
@@ -36,188 +24,79 @@ let substituteTemplateValues = s =>
...
+    let duneProjectFile = Path.(projectPath / "dune-project");
```
Comment (ulrikstrid) [inline npm-cli/src/Pesy.re]: I decided to keep generating a dune-project file and the .opam file, what do you think?

File: npm-cli/src/Pesy.re (hunk 3 above)
Comment (ulrikstrid) [reply npm-cli/src/Pesy.re]: We now read the `--template=` argument

File: npm-cli/.npmignore (hunk 1 above)
Comment (prometheansacrifice) [reply npm-cli/.npmignore]: Yeah. It wouldn't hurt to publish the .re files to npm.

File: npm-cli/src/Pesy.re [hunk 6]
```diff
@@ -36,188 +24,99 @@ let substituteTemplateValues = s =>
...
+        files,
+        file => {
+          let () = readFile(file)->substituteTemplateValues |> write(file);
+
+          renameSync(file, substituteTemplateValues(file));
+        },
+      );
+
+      let libKebab = packageNameKebabSansScope;
+      let duneProjectFile = Path.(projectPath / "dune-project");
```
Comment (prometheansacrifice) [inline npm-cli/src/Pesy.re]: Isn't `dune-project` file a part of the hello-reason-pesy repo?

File: npm-cli/src/Bindings.re [hunk 7]
```diff
@@ -1,6 +1,21 @@
...
+
 [@bs.val] [@bs.module "process"] external cwd: unit => string = "cwd";
+[@bs.val] [@bs.module "process"] external argv: array(string) = "argv";
 [@bs.val] external scriptPath: string = "__dirname";
 [@bs.val] [@bs.module "child_process"]
 external exec: (string, (Js.Nullable.t(_), string, string) => unit) => unit =
   "";
+
+[@bs.module]
+external download_git: (string, string, option('a) => unit) => unit =
```
Comment (prometheansacrifice) [inline npm-cli/src/Bindings.re]: Can we do something more light-weight? Downloading the git history seems unnecessary. We can work with this for the time being  though 👍

Comment (ulrikstrid) [reply npm-cli/src/Bindings.re]: It doesn't clone by default so it should just download the repo as a zip and then extract it. We could probably implement this ontop of the `download` package ourselves but this package allows us to easily download from gitlab and bitbucket as well.

File: npm-cli/src/Pesy.re (hunk 6 above)
Comment (ulrikstrid) [reply npm-cli/src/Pesy.re]: It is, but in this case we're generating a `dune-project` and `package-name.opam`if the template doesn't include them for some reason.
But I could just remove these, in that case we could potentially support bucklescript as well if it works with `esy` in the future.

File: npm-cli/.npmignore (hunk 1 above)
Comment (ulrikstrid) [reply npm-cli/.npmignore]: Okay, I can just remove this from the ignore again then. They are not used in this PR anymore but I guess they don't hurt either?

**Comment by prometheansacrifice** [comment]
I guess this will address #5 too

File: npm-cli/src/Pesy.re [hunk 8]
```diff
@@ -36,188 +24,94 @@ let substituteTemplateValues = s =>
...
-/*       "require: [%s]", */
-/*       String.concat(", ", [spf("\"%s\"", packageLibName)]), */
-/*     ), */
-/*   ], */
-/* ]); */
-/* print_newline(); */
+let substituteFileNames = s =>
+  s
+  |> Js.String.replaceByRe(
+       [%bs.re "/__PACKAGE_NAME_FULL__/g"],
```
Comment (prometheansacrifice) [inline npm-cli/src/Pesy.re]: Why was this change in the template variable required? (`__PACKAGE_NAME_FULL__`)

File: npm-cli/src/Pesy.re (hunk 6 above)
Comment (prometheansacrifice) [reply npm-cli/src/Pesy.re]: Well, then this brings up the issue of validation of the templates. For now let's track this in a separate issue and assume the template has all the files needed.

File: npm-cli/src/Bindings.re (hunk 7 above)
Comment (prometheansacrifice) [reply npm-cli/src/Bindings.re]: Oh, I misunderstood then. This is great!

File: npm-cli/src/Pesy.re (hunk 6 above)
Comment (prometheansacrifice) [reply npm-cli/src/Pesy.re]: https://github.com/esy/pesy/issues/38

File: npm-cli/src/Pesy.re (hunk 8 above)
Comment (ulrikstrid) [reply npm-cli/src/Pesy.re]: Windows couldn't handle files named `<PACKAGE_NAME_FULL>` so this is specific for file names, the in-file replaces are the same as before.

**Comment by ulrikstrid** [comment]
The template variable change is only for the file names, it's the same inside the files.

**Pull request merged**